### DISCUSSION TITLE
How can I check `null!` or `null` in property initializer?

### DISCUSSION THREAD

**Discussion by kyoyama-kazusa** [dotnet/roslyn#51358] [Q&A] [answered]
Hello! I'm stuck in checking property initializer.

I want to check the properties with `null!` or `null` as the property initializer:

```csharp
public Class? Property1 { get; }
public Class Property2 { get; } = null!;
public Class? Property3 { get; } = null;
```

As the example shown, properties `Property2` and `Property3` should be detected as I expected. How can I write code using Roslyn to check them?

Thanks.

**Comment by bernd5** [comment]
You mean that you want to find properties explicitly initialized with `null` or `null!`?
Why?
Property1 is null, too.

**Comment by PathogenDavid** [comment, marked as answer]
The expression for `Property2` will be wrapped in a `SuppressNullableWarningExpression`, whereas `Property3` will not.

[SharpLab has a nice visualization of the syntax tree](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzhgHwGIA7AVwBsLtgKYACGEmugWACgOABAZnq4BM9AMLVcuDgG8O9Wfz6i8uAPz0AClAgAHGFAwBPAIz1J9AOYwMAbnoBfGXN4ixudZp179Q0xet36ALz05FQAhFYOsk6K4qoa2roGfD6WNraBwZQUEey2QA):

![image](https://user-images.githubusercontent.com/278957/108614191-4693bd80-73be-11eb-961b-913b6d7afd95.png)

**Reply by kyoyama-kazusa** [reply]
That's awesome! Thank you :D

**Reply by PathogenDavid** [reply]
No problem, glad I could help!

**Reply by PathogenDavid** [reply]
BTW, I forgot to mention [a similar tool exists within Visual Studio](https://docs.microsoft.com/en-us/dotnet/csharp/roslyn-sdk/syntax-visualizer?tabs=csharp) as well.
